File: benchmark/punctuation-restoration/parity/pcs47-probe-layers.mjs
// onnxruntime-web side of parity/pcs47-probe-layers.py: run the augmented graphs on WASM (1 thread), compare
// every exposed tensor with the native run, in graph order.
// usage: node parity/pcs47-probe-layers.mjs int8 fp32
import { readFile, writeFile } from 'node:fs/promises';
import { ort } from '/home/jiangzhuo/Desktop/kizunaai/sokuji/.claude/worktrees/research-asr-punctuation/benchmark/punctuation-restoration/lib/node-env.mjs';

const OUT = '/home/jiangzhuo/.cache/sokuji-punct-bench/pcs47-probe';
const man = JSON.parse(await readFile(`${OUT}/manifest.json`, 'utf8'));
ort.env.wasm.numThreads = 1;
const short = (n) => n.replace('/bert_model/', '').replace('/encoder/', '').replace(/_output_0.*$/, '');
const summary = {};

for (const tag of process.argv.slice(2)) {
  const m = man.models[tag];
  const session = await ort.InferenceSession.create(new Uint8Array(await readFile(`${OUT}/${m.model}`)), { executionProviders: ['wasm'] });
  summary[tag] = {};
  for (const [k, ids] of Object.entries(man.inputs)) {
    const flat = ids[0];
    const r = await session.run({ input_ids: new ort.Tensor('int64', BigInt64Array.from(flat.map(BigInt)), [1, flat.length]) });
    console.log(`\n== ${tag} ${k} (${flat.length} ids)`);
    const rows = [];
    for (const e of m.entries.filter((x) => x.input === k)) {
      const bytes = new Uint8Array(await readFile(`${OUT}/${e.file}`));
      const got = r[e.name].data;
      if (e.kind === 'xq') {
        let n = 0;
        for (let i = 0; i < bytes.length; i++) if (got[i] !== bytes[i]) n++;
        rows.push({ kind: 'xq', name: short(e.name), differ: n, of: bytes.length });
        console.log(`xq     ${short(e.name).padEnd(56)} differs ${n} / ${bytes.length}`);
      } else {
        const want = new Float32Array(bytes.buffer);
        let maxAbs = 0;
        let maxVal = 0;
        let n = 0;
        for (let i = 0; i < want.length; i++) {
          const d = Math.abs(got[i] - want[i]);
          if (d > 0) n++;
          if (d > maxAbs) maxAbs = d;
          if (Math.abs(want[i]) > maxVal) maxVal = Math.abs(want[i]);
        }
        rows.push({ kind: e.kind, name: short(e.name), differ: n, of: want.length, maxAbs, maxVal });
        console.log(`${e.kind.padEnd(6)} ${short(e.name).padEnd(56)} differs ${n} / ${want.length}  max |diff| ${maxAbs.toExponential(2)}  (max |value| ${maxVal.toExponential(2)})`);
      }
    }
    summary[tag][k] = rows;
  }
  await session.release();
}
await writeFile(`${OUT}/comparison.json`, JSON.stringify(summary, null, 1));
console.log(`\nwrote ${OUT}/comparison.json`);

File: benchmark/punctuation-restoration/parity/pcs47-probe-layers.py
"""Where do onnxruntime-web and native onnxruntime part ways on PCS-47?

Expose every LayerNormalization output (fp32 and int8 graphs) and both DynamicQuantizeLinear
outputs that carry data (quantized tensor, scale; int8 graph) as graph outputs, run two real
128-token windows natively, and save every array for parity/pcs47-probe-layers.mjs (onnxruntime-web).
"""
import json
import os

import numpy as np
import onnx
import onnxruntime as ort
from onnx import TensorProto, helper

D = "/home/jiangzhuo/.cache/sokuji-punct-bench/pcs47"
OUT = "/home/jiangzhuo/.cache/sokuji-punct-bench/pcs47-probe"
REF = "/home/jiangzhuo/Desktop/kizunaai/sokuji/.claude/worktrees/research-asr-punctuation/benchmark/punctuation-restoration/results/parity-pcs47-ref-fp32.json"
os.makedirs(OUT, exist_ok=True)

with open(REF, encoding="utf-8") as fh:
    rows = {r["id"]: r for r in json.load(fh)["rows"]}
inputs = {k: np.array([[0] + rows[f"{k}-long"]["sp_ids_pre"][:128] + [2]], dtype=np.int64) for k in ["en", "zh"]}


def augment(src, tag):
    m = onnx.load(src)
    have = {o.name for o in m.graph.output}
    extra = []
    for n in m.graph.node:
        if n.op_type == "LayerNormalization":
            extra.append((n.output[0], TensorProto.FLOAT, "ln"))
        elif n.op_type == "DynamicQuantizeLinear":
            extra.append((n.output[0], TensorProto.UINT8, "xq"))
            extra.append((n.output[1], TensorProto.FLOAT, "scale"))
    extra = [e for e in extra if e[0] not in have]
    for name, elem, _ in extra:
        m.graph.output.append(helper.make_tensor_value_info(name, elem, None))
    path = f"{OUT}/{tag}.onnx"
    onnx.save(m, path)
    return path, extra


models = {}
for tag, src in [("int8", f"{D}/model.int8.onnx"), ("fp32", f"{D}/model.onnx")]:
    path, extra = augment(src, tag)
    so = ort.SessionOptions()
    so.intra_op_num_threads = 2
    s = ort.InferenceSession(path, sess_options=so, providers=["CPUExecutionProvider"])
    names = [e[0] for e in extra]
    entries = []
    for k, x in inputs.items():
        vals = s.run(names, {"input_ids": x})
        for i, ((name, _, kind), v) in enumerate(zip(extra, vals)):
            f = f"{tag}-{k}-{i}.bin"
            np.ascontiguousarray(v).tofile(f"{OUT}/{f}")
            entries.append({"input": k, "name": name, "kind": kind, "dtype": str(v.dtype), "shape": list(v.shape), "file": f})
    models[tag] = {"model": os.path.basename(path), "entries": entries}
    print(tag, "exposed", len(extra), "tensors ->", path, os.path.getsize(path), "bytes", flush=True)

with open(f"{OUT}/manifest.json", "w") as fh:
    json.dump(
        {"inputs": {k: x.tolist() for k, x in inputs.items()}, "models": models, "ort": ort.__version__},
        fh,
    )
print("wrote", f"{OUT}/manifest.json")
